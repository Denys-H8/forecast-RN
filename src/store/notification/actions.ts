import { FirebaseMessagingTypes } from '@react-native-firebase/messaging';
import * as ACTION from './actionTypes';

export const notification = (routeName: string) => ({ type: ACTION.NOTIFICATION, payload: routeName });
export const addNotification = (notification: FirebaseMessagingTypes.RemoteMessage) => ({
  type: ACTION.ADD_NOTIFICATION,
  payload: notification,
});
export const removeNotification = (notification: FirebaseMessagingTypes.RemoteMessage) => ({
  type: ACTION.REMOVE_NOTIFICATION,
  payload: notification,
});
