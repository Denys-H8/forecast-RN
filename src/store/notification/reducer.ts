import { FirebaseMessagingTypes } from '@react-native-firebase/messaging';
import * as ACTION from './actionTypes';

type Props = {
  routeName: string;
  notification: FirebaseMessagingTypes.RemoteMessage | null;
};

export const initialState: Props = {
  routeName: '',
  notification: null,
};

export const notificationReducer = (state = initialState, action: { type: string; payload?: any }) => {
  switch (action.type) {
    case ACTION.ADD_NOTIFICATION: {
      return {
        ...state,
        notification: action.payload,
      };
    }
    case ACTION.REMOVE_NOTIFICATION: {
      return {
        ...state,
        notification: null,
      };
    }

    default:
      return state;
  }
};
