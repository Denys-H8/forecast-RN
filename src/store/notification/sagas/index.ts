import watchNotificationAsync from './notificationSaga';

export default watchNotificationAsync;
