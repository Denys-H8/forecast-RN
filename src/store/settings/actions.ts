import * as ACTION from './actionTypes';

export const changedName = (name: string) => ({ type: ACTION.CHANGED_NAME, payload: name });
export const changedAge = (age: string) => ({ type: ACTION.CHANGED_AGE, payload: age });
export const changedDayPeriod = (days: number) => ({ type: ACTION.CHANGED_DAY_PERIOD, payload: days });
export const changedUpdateTime = (time: number) => ({ type: ACTION.CHANGED_UPDATE_TIME, payload: time });
export const changedTemperatureScale = (scale: string) => ({
  type: ACTION.CHANGED_TEMPERATURE_SCALE,
  payload: scale,
});
export const setTemperatureScale = (scale: string) => ({
  type: ACTION.SET_TEMPERATURE_SCALE,
  payload: scale,
});
