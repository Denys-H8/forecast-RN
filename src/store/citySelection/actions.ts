import * as ACTION from './actionTypes';
import { TGroupWeather, TData, TCity } from './reducer';

export const queryChanged = (query: string, data: TData[]) => ({
  type: ACTION.QUERY_CHANGED,
  payload: { query, data },
});
export const getWeatherForCities = () => ({ type: ACTION.GET_WEATHER_FOR_CITIES });
export const getWeatherForCitiesSucess = (data: TGroupWeather) => ({
  type: ACTION.GET_WEATHER_FOR_CITIES_SUCESS,
  payload: data,
});
export const resetQuery = () => ({ type: ACTION.RESET_QUERY });
export const changeCity = (city: TCity) => ({
  type: ACTION.CHANGE_CITY,
  payload: city,
});
export const pressedCity = (city: TCity) => ({
  type: ACTION.PRESSED_CITY,
  payload: city,
});
export const setLoading = (value: boolean) => ({ type: ACTION.SET_LOADING, payload: value });
