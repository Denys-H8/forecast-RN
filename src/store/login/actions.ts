import * as ACTION from './actionTypes';
import { TUser } from './reducer';

export const login = () => ({ type: ACTION.LOGIN });
export const logout = () => ({ type: ACTION.LOGOUT });
export const logoutSuccess = () => ({ type: ACTION.LOGOUT_SUCCESS });
export const loginStarted = () => ({ type: ACTION.LOGIN_STARTED });
export const loginSuccess = (user: TUser) => ({ type: ACTION.LOGIN_SUCCESS, payload: user });
export const changedEmail = (email: string | false) => ({ type: ACTION.CHANGED_EMAIL, payload: email });
export const changedPassword = (password: string) => ({ type: ACTION.CHANGED_PASSWORD, payload: password });
export const failedEmail = () => ({ type: ACTION.FAILED_EMAIL });
export const failedPassword = () => ({ type: ACTION.FAILED_PASSWORD });
export const startLoading = () => ({ type: ACTION.START_LOADING });
export const loading = (value: boolean) => ({ type: ACTION.LOADING_LOGIN, payload: value });
