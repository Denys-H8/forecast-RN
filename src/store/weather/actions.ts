import * as ACTION from './actionTypes';
import { TWeather, TWeatherStats } from './reducer';

export const getWeather = () => ({ type: ACTION.GET_WEATHER });
export const getWeatherSuccess = (weatherData: TWeatherData) => ({
  type: ACTION.GET_WEATHER_SUCCESS,
  payload: weatherData,
});
export const loading = (value: boolean) => ({ type: ACTION.LOADING, payload: value });
export const lateNavigation = (routeName: string | undefined) => ({ type: ACTION.LATE_NAVIGATION, payload: routeName });
export const resetLateNavigation = () => ({ type: ACTION.RESET_LATE_NAVIGATION });

export type TWeatherData = {
  weather: TWeather;
  weatherStats: TWeatherStats;
};
