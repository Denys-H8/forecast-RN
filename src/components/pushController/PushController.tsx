import React, { useEffect } from 'react';
import { notificationActions, weatherActions } from 'src/store/rootActions';
import messaging, { FirebaseMessagingTypes } from '@react-native-firebase/messaging';
import { useDispatch, useSelector } from 'react-redux';
import { userSelector } from 'src/store/login/selectors';
import { View } from 'react-native';
import { notificationSelector } from 'src/store/notification/selectors';
import NotificationCard from './components/notificationCard';
import { styles } from './pushController.styles';

const PushController: React.FC = ({ children }) => {
  const dispatch = useDispatch();
  const user = useSelector(userSelector);
  const notification = useSelector(notificationSelector);

  useEffect(() => {
    handleNotifications();
  }, []);

  const handleNotifications = () => {
    messaging().setBackgroundMessageHandler(async (remoteMessage) => {
      console.log('Message is handed in the background!', remoteMessage);
    });

    messaging().onMessage(async (remoteMessage: FirebaseMessagingTypes.RemoteMessage) => {
      dispatch(notificationActions.addNotification(remoteMessage));
    });

    messaging().onNotificationOpenedApp((remoteMessage: FirebaseMessagingTypes.RemoteMessage) => {
      if (user && remoteMessage?.data?.type) {
        dispatch(notificationActions.notification(remoteMessage?.data?.type));
      } else {
        dispatch(weatherActions.lateNavigation(remoteMessage?.data?.type));
      }
    });

    messaging()
      .getInitialNotification()
      .then((remoteMessage: FirebaseMessagingTypes.RemoteMessage | null) => {
        dispatch(weatherActions.lateNavigation(remoteMessage?.data?.type));
      });
  };
  return (
    <View style={styles.app}>
      {children}
      <View style={styles.container}>
        {notification && <NotificationCard data={notification} key={notification.messageId} />}
      </View>
    </View>
  );
};

export default PushController;
