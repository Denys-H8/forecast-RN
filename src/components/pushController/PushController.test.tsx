import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import PushController from './PushController';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { View } from 'react-native';
import { notification } from 'src/store/notification/actions';

const mockStore = configureStore([]);

const notificationValue = {
  data: {
    type: 'Weather',
  },
  notification: {
    title: 'Some title',
    body: 'Some message',
  },
};
const store = mockStore({
  loginReducer: {
    user: { email: 'root@gmail.com' },
  },
  notificationReducer: {
    notification: notificationValue,
  },
});

const storeWithoutNotification = mockStore({
  loginReducer: {
    user: { email: 'root@gmail.com' },
  },
  notificationReducer: {
    notification: null,
  },
});

describe('PushController ', () => {
  test('should render with notification', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <PushController>
          <View style={{ backgroundColor: 'green' }} />
        </PushController>
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render without notification', () => {
    const { toJSON } = render(
      <Provider store={storeWithoutNotification}>
        <PushController>
          <View style={{ backgroundColor: 'green' }} />
        </PushController>
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  // test('should handle onMessage', () => {
  //   store.dispatch = jest.fn();

  //   const { toJSON } = render(
  //     <Provider store={store}>
  //       <PushController>
  //         <View style={{ backgroundColor: 'green' }} />
  //       </PushController>
  //     </Provider>,
  //   );

  //   expect(store.dispatch).toHaveBeenCalledTimes(1);
  //   expect(store.dispatch).toHaveBeenCalledWith(notification(notificationValue.data.type));
  // });
});
