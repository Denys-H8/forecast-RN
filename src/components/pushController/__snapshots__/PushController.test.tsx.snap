// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PushController  should render with notification 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "green",
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#191C39",
        "bottom": undefined,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#464960",
          "opacity": 0.8,
        }
      }
      testID="NotificationCard.TouchableOpacity"
    >
      <View
        style={
          Object {
            "backgroundColor": "#7B61FF",
            "height": 5,
          }
        }
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 30,
            "paddingVertical": 15,
          }
        }
      >
        <IconMock
          height={28}
          width={28}
        />
        <View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "WorkSans-ExtraBold",
                  "fontSize": 52,
                  "lineHeight": 61,
                },
                Object {
                  "color": "#fff",
                  "fontSize": 18,
                  "lineHeight": 20,
                  "marginBottom": 8,
                },
              ]
            }
          >
            Some title
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "WorkSans-ExtraBold",
                  "fontSize": 52,
                  "lineHeight": 61,
                },
                Object {
                  "color": "#fff",
                  "fontSize": 16,
                  "lineHeight": 16,
                },
              ]
            }
          >
            Some message
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 5,
            }
          }
          testID="NotificationCard.TouchableOpacity.Remove"
        >
          <IconMock
            height={24}
            width={24}
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`PushController  should render without notification 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "green",
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#191C39",
        "bottom": undefined,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 20,
      }
    }
  />
</View>
`;
